cpp/src/IceLocatorDiscovery/PluginI.cpp

```diff
@@ -52,40 +51,43 @@ class Request : public std::enable_shared_from_this<Request>
     const Ice::OperationMode _mode;
     const Ice::Context _context;
     const Ice::ByteSeq _inParams;
-    AMDCallback _amdCB;
+    function<void(bool, const pair<const uint8_t*, const uint8_t*>)> _responseCallback;
+    function<void(exception_ptr)> _exceptionCallback;
     exception_ptr _exception;
 
-    Ice::LocatorPrxPtr _locatorPrx;
+    optional<Ice::LocatorPrx> _locatorPrx;
 };
 using RequestPtr = std::shared_ptr<Request>;
 
-class LocatorI : public Ice::BlobjectArrayAsync,
-                 public IceUtil::TimerTask,
-                 public std::enable_shared_from_this<LocatorI>
+class LocatorI final :
+    public Ice::BlobjectArrayAsync,
+    public IceUtil::TimerTask,
+    public std::enable_shared_from_this<LocatorI>
 {
 public:
 
-    LocatorI(const string&, const LookupPrxPtr&, const Ice::PropertiesPtr&, const string&, const Ice::LocatorPrxPtr&);
-    void setLookupReply(const LookupReplyPrxPtr&);
+    LocatorI(const string&, const LookupPrx&, const Ice::PropertiesPtr&, const string&, const Ice::LocatorPrx&);
+    void setLookupReply(const LookupReplyPrx&);
 
-    virtual void ice_invokeAsync(pair<const uint8_t*, const uint8_t*>,
-                                 function<void(bool, const pair<const uint8_t*, const uint8_t*>&)>,
-                                 function<void(exception_ptr)>,
-                                 const Ice::Current&);
+    void ice_invokeAsync(
+        pair<const uint8_t*, const uint8_t*>,
+        function<void(bool, const pair<const uint8_t*, const uint8_t*>&)>,
+        function<void(exception_ptr)>,
+        const Ice::Current&) final;
 
     void foundLocator(const Ice::LocatorPrxPtr&);
     void invoke(const Ice::LocatorPrxPtr&, const RequestPtr&);
 
-    vector<Ice::LocatorPrxPtr> getLocators(const string&, const chrono::milliseconds&);
+    vector<Ice::LocatorPrx> getLocators(const string&, const chrono::milliseconds&);
 
     void exception(std::exception_ptr);
 
 private:
 
-    virtual void runTimerTask();
+    void runTimerTask() final;
 
     LookupPrxPtr _lookup;
-    vector<pair<LookupPrxPtr, LookupReplyPrxPtr> > _lookups;
+    vector<pair<LookupPrx, optional<LookupReplyPrx>> > _lookups;
```

Comment:
Maybe we can initialize `_lookups` in `setLookupReply` and avoid using an optional proxy for the reply.